simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/ChangeAvailabilityRequestHandler.java

```diff
@@ -0,0 +1,112 @@
+     *
+     * @param stationState         {@link StationState}
+     * @param stationMessageSender {@link StationMessageSender}
+     */
+    public ChangeAvailabilityRequestHandler(StationState stationState, StationMessageSender stationMessageSender) {
+        this(stationState, stationMessageSender, DEFAULT_NUMBER_OF_ATTEMPTS, DELAY_BETWEEN_ATTEMPTS);
+    }
+
+    /**
+     * Create an instance.
+     *
+     * @param stationState                 {@link StationState}
+     * @param stationMessageSender         {@link StationMessageSender}
+     * @param attemptsInMillis             number of attempts to try to change the state before giving up
+     * @param delayBetweenAttemptsInMillis delay between attempts
+     */
+    public ChangeAvailabilityRequestHandler(StationState stationState, StationMessageSender stationMessageSender,
+                                            int attemptsInMillis, int delayBetweenAttemptsInMillis) {
+        this.stationState = stationState;
+        this.stationMessageSender = stationMessageSender;
+        this.attemptsInMillis = attemptsInMillis;
+        this.delayBetweenAttemptsInMillis = delayBetweenAttemptsInMillis;
+    }
+
+    /**
+     * Handle {@link ChangeAvailabilityRequest} request.
+     * It has 4 scenarios:
+     * 1. Send response with ACCEPTED status when EVSE state is the same as requested.
+     * 2. Change EVSE state to the requested state when they do not match.
+     * In addition send response with ACCEPTED status and StatusNotification request for every EVSE Connector.
+     * 3. When a transaction is in progress.
+     * Send response with SCHEDULED status and schedule a task to change the EVSE state when transaction is finished.
+     * 4. Send response with REJECTED status if exception occurs.
+     *
+     * @param callId  identity of the message
+     * @param request incoming request from the server
+     */
+    @Override
+    public void handle(String callId, ChangeAvailabilityRequest request) {
+        try {
+            Evse evse = stationState.findEvse(request.getEvseId());
+
+            EvseState requestedEvseState = AvailabilityStateMapper.mapFrom(request.getOperationalStatus());
+
+            if (evse.hasOngoingTransaction()) {
+
+                changeAvailabilityFuture.runAsync(evse, requestedEvseState, attemptsInMillis, delayBetweenAttemptsInMillis);
+
+                sendResponseWithStatus(callId, SCHEDULED);
+
+            } else {
+
+                sendResponseWithStatus(callId, ACCEPTED);
+
+                if (!evse.hasState(requestedEvseState)) {
+
+                    evse.setEvseState(requestedEvseState);
+
+                    // for every connector send StatusNotification request
+                    for (Connector connector : evse.getConnectors()) {
+                        stationMessageSender.sendStatusNotification(evse, connector);
+                    }
+
+                }
+
+            }
+
+        } catch (Exception e) {
```

Comment:
We already have try-catch on the top level, in consumer, which should reply with CALL_ERROR in case of unexpected errors. We should reply with rejected in case of business error, e.g. provided evseId is not present on the station.

Reply:
Spec does not say anything about error handling. But the response might contain status REJECTED. So apparently if there is an exception we can send it. Is there where I can read about business errors?

Reply:
Usually use-cases from spec specify conditions when rejected is appropriate to send, although I couldn't find description for this particular message. In general if call couldn't be processed we send call error with some description (for ex. message from the exception), reject means that everything went well, but certain conditions don't allow to accept the request. Besides, how do you know that exception wasn't thrown after evse was scheduled for disabling?

Reply:
I will remove this try...catch and sending of REJECTED status. As rejected does not include any description of what went wrong.